Issue with virtual:astro:manifest build

### Astro Info

```block
Astro                    v6.0.6
Vite                     v7.3.1
Node                     v25.2.1
System                   Windows (x64)
Package Manager          pnpm
Output                   server
Adapter                  @astrojs/vercel (v10.0.1)
Integrations             @astrojs/sitemap (v3.7.1)
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When doing a pnpm build, it terminates with an error where rollup failed to resolve import virtual:astro routes from virtual:astro:manifest. I was getting this in vercel, but also in local.

```
pnpm run build

> lioncommunity-web@1.1.0 build D:\Users\XXXXXXX\Documents\Github\project-web
> astro build

14:17:12 [vite] Re-optimizing dependencies because lockfile has changed
14:17:12 [types] Generated 149ms
14:17:12 [build] output: "server"
14:17:12 [build] mode: "server"
14:17:12 [build] directory: D:\Users\XXXXXXX\Documents\Github\project-web\dist\
14:17:12 [build] adapter: @astrojs/vercel
14:17:12 [build] Collecting build info...
14:17:12 [build] ✓ Completed in 198ms.
14:17:12 [build] Building server entrypoints...
14:17:13 [vite] ✓ built in 1.21s
14:17:15 [vite] ✓ built in 1.94s
14:17:15 [ERROR] [vite] ✗ Build failed in 228ms
[vite]: Rollup failed to resolve import "virtual:astro:routes" from "virtual:astro:manifest".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
  Location:
    D:\Users\XXXXXXX\Documents\Github\project-web\node_modules\.pnpm\vite@7.3.1_@types+node@25.0_1f92fa69447a5c463e82235c2a65bf5a\node_modules\vite\dist\node\chunks\config.js:33635:57
  Stack trace:
    at viteLog (file:///D:/Users/XXXXXXX/Documents/Github/project-web/node_modules/.pnpm/vite@7.3.1_@types+node@25.0_1f92fa69447a5c463e82235c2a65bf5a/node_modules/vite/dist/node/chunks/config.js:33635:57)
    at onLog (file:///D:/Users/XXXXXXX/Documents/Github/project-web/node_modules/.pnpm/vite@7.3.1_@types+node@25.0_1f92fa69447a5c463e82235c2a65bf5a/node_modules/vite/dist/node/chunks/config.js:33467:4)
    at Object.logger [as onLog] (file:///D:/Users/XXXXXXX/Documents/Github/project-web/node_modules/.pnpm/rollup@4.59.0/node_modules/rollup/dist/es/shared/node-entry.js:22968:9)
    at file:///D:/Users/XXXXXXX/Documents/Github/project-web/node_modules/.pnpm/rollup@4.59.0/node_modules/rollup/dist/es/shared/node-entry.js:21670:26
 ELIFECYCLE  Command failed with exit code 1.
```

In dev

```
14:30:11 [ERROR] [vite] Internal server error: Failed to resolve import "virtual:astro:routes" from "virtual:astro:manifest". Does the file exist?
  Plugin: vite:import-analysis
  File: virtual:astro:manifest:4:29
  2  |  					import { deserializeManifest as _deserializeManifest } from 'astro/app';
  3  |  					import { renderers } from 'virtual:astro:renderers';
  4  |  					import { routes } from 'virtual:astro:routes';
     |                               ^
  5  |  					import { pageMap } from 'virtual:astro:pages';
```

When donwgrading to astro 5the issue disappears

### What's the expected result?

The expected result is to build correctly

### Link to Minimal Reproducible Example

I cannot reproduceit

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/lucky-teams-play.md
+++ b/.changeset/lucky-teams-play.md
@@ -1,0 +1,5 @@
+---
+'astro': patch
+---
+
+Fixes a build error where using `astro:config/client` inside a `<script>` tag would cause Rollup to fail with "failed to resolve import `virtual:astro:routes` from `virtual:astro:manifest`"

--- a/packages/astro/src/core/create-vite.ts
+++ b/packages/astro/src/core/create-vite.ts
@@ -136,7 +136,7 @@ export async function createVite(
 			}),
 			vitePluginStaticPaths(),
 			await astroPluginRoutes({ routesList, settings, logger, fsMod: fs, command }),
-			astroVirtualManifestPlugin(),
+			astroVirtualManifestPlugin({ settings }),
 			vitePluginEnvironment({ settings, astroPkgsConfig, command }),
 			pluginPage({ routesList }),
 			pluginPages({ routesList }),

--- a/packages/astro/src/manifest/serialized.ts
+++ b/packages/astro/src/manifest/serialized.ts
@@ -72,6 +72,16 @@ export function serializedManifestPlugin({
 			server.watcher.on('change', (path) => reloadManifest(path, server));
 		},
 
+		// Restrict to server environments only since the generated code imports
+		// server-only virtual modules (virtual:astro:routes, virtual:astro:pages)
+		applyToEnvironment(environment) {
+			return (
+				environment.name === ASTRO_VITE_ENVIRONMENT_NAMES.astro ||
+				environment.name === ASTRO_VITE_ENVIRONMENT_NAMES.ssr ||
+				environment.name === ASTRO_VITE_ENVIRONMENT_NAMES.prerender
+			);
+		},
+
 		resolveId: {
 			filter: {
 				id: new RegExp(`^${SERIALIZED_MANIFEST_ID}$`),

--- a/packages/astro/src/manifest/virtual-module.ts
+++ b/packages/astro/src/manifest/virtual-module.ts
@@ -2,13 +2,59 @@ import type { Plugin } from 'vite';
 import { AstroError, AstroErrorData } from '../core/errors/index.js';
 import { SERIALIZED_MANIFEST_ID } from './serialized.js';
 import { ASTRO_VITE_ENVIRONMENT_NAMES } from '../core/constants.js';
+import { fromRoutingStrategy, toFallbackType, toRoutingStrategy } from '../core/app/common.js';
+import type { AstroSettings } from '../types/astro.js';
 
 const VIRTUAL_SERVER_ID = 'astro:config/server';
 const RESOLVED_VIRTUAL_SERVER_ID = '\0' + VIRTUAL_SERVER_ID;
 const VIRTUAL_CLIENT_ID = 'astro:config/client';
 const RESOLVED_VIRTUAL_CLIENT_ID = '\0' + VIRTUAL_CLIENT_ID;
 
-export default function virtualModulePlugin(): Plugin {
+export default function virtualModulePlugin({ settings }: { settings: AstroSettings }): Plugin {
+	// Pre-compute the client config values from settings so that astro:config/client
+	// doesn't need to import from virtual:astro:manifest (which pulls in server-only
+	// virtual modules like virtual:astro:routes and virtual:astro:pages that are
+	// restricted to server environments via applyToEnvironment).
+	const config = settings.config;
+
+	let i18nCode = 'const i18n = undefined;';
+	if (config.i18n) {
+		// Apply the same toRoutingStrategy → fromRoutingStrategy roundtrip that the
+		// serialized manifest uses, to ensure consistent routing config values.
+		const strategy = toRoutingStrategy(config.i18n.routing, config.i18n.domains);
+		const fallbackType = toFallbackType(config.i18n.routing);
+		const routing = fromRoutingStrategy(strategy, fallbackType);
+		i18nCode = `const i18n = {
+  defaultLocale: ${JSON.stringify(config.i18n.defaultLocale)},
+  locales: ${JSON.stringify(config.i18n.locales)},
+  routing: ${JSON.stringify(routing)},
+  fallback: ${JSON.stringify(config.i18n.fallback)}
+};`;
+	}
+
+	let imageCode = 'const image = undefined;';
+	if (config.image) {
+		imageCode = `const image = {
+  objectFit: ${JSON.stringify(config.image.objectFit)},
+  objectPosition: ${JSON.stringify(config.image.objectPosition)},
+  layout: ${JSON.stringify(config.image.layout)},
+};`;
+	}
+
+	const clientConfigCode = `
+${i18nCode}
+${imageCode}
+const base = ${JSON.stringify(config.base)};
+const trailingSlash = ${JSON.stringify(config.trailingSlash)};
+const site = ${JSON.stringify(config.site)};
+const compressHTML = ${JSON.stringify(config.compressHTML)};
+const build = {
+  format: ${JSON.stringify(config.build.format)},
+};
+
+export { base, i18n, trailingSlash, site, compressHTML, build, image };
+`;
+
 	return {
 		name: 'astro-manifest-plugin',
 		resolveId: {
@@ -30,41 +76,11 @@ export default function virtualModulePlugin(): Plugin {
 			},
 			handler(id) {
 				if (id === RESOLVED_VIRTUAL_CLIENT_ID) {
-					// There's nothing wrong about using `/client` on the server
-					const code = `
-import { manifest } from '${SERIALIZED_MANIFEST_ID}'
-import { fromRoutingStrategy } from 'astro/app';
-
-let i18n = undefined;
-if (manifest.i18n) {
-i18n = {
-  defaultLocale: manifest.i18n.defaultLocale,
-  locales: manifest.i18n.locales,
-  routing: fromRoutingStrategy(manifest.i18n.strategy, manifest.i18n.fallbackType),
-  fallback: manifest.i18n.fallback
-  };
-}
-
-let image = undefined;
-if (manifest.image) {
-  image = {
-    objectFit: manifest.image.objectFit,
-    objectPosition: manifest.image.objectPosition,
-    layout: manifest.image.layout,
-  };
-}
-
-const base = manifest.base;
-const trailingSlash = manifest.trailingSlash;
-const site = manifest.site;
-const compressHTML = manifest.compressHTML;
-const build = {
-  format: manifest.buildFormat,
-};
-
-export { base, i18n, trailingSlash, site, compressHTML, build, image };
-				`;
-					return { code };
+					// astro:config/client inlines values directly from settings instead of
+					// importing from virtual:astro:manifest to avoid pulling server-only
+					// virtual modules (virtual:astro:routes, virtual:astro:pages) into the
+					// client environment where they are not available.
+					return { code: clientConfigCode };
 				}
 				if (id === RESOLVED_VIRTUAL_SERVER_ID) {
 					if (this.environment.name === ASTRO_VITE_ENVIRONMENT_NAMES.client) {

--- a/packages/astro/test/fixtures/astro-manifest-client-script/astro.config.mjs
+++ b/packages/astro/test/fixtures/astro-manifest-client-script/astro.config.mjs
@@ -1,0 +1,6 @@
+import { defineConfig } from "astro/config";
+
+// https://astro.build/config
+export default defineConfig({
+	site: "https://example.com",
+});

--- a/packages/astro/test/fixtures/astro-manifest-client-script/package.json
+++ b/packages/astro/test/fixtures/astro-manifest-client-script/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "@test/astro-manifest-client-script",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "astro": "workspace:*"
+  }
+}

--- a/packages/astro/test/fixtures/astro-manifest-client-script/src/pages/index.astro
+++ b/packages/astro/test/fixtures/astro-manifest-client-script/src/pages/index.astro
@@ -1,0 +1,18 @@
+---
+---
+<!DOCTYPE html>
+<html lang="en">
+<head>
+	<meta charset="UTF-8"/>
+	<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
+	<title>Document</title>
+</head>
+<body>
+	<h1>Hello, World!</h1>
+	<p id="base"></p>
+	<script>
+		import { base } from "astro:config/client";
+		document.getElementById('base').textContent = base;
+	</script>
+</body>
+</html>

--- a/packages/astro/test/serializeManifest.test.js
+++ b/packages/astro/test/serializeManifest.test.js
@@ -89,6 +89,26 @@ describe('astro:config/client', () => {
 	});
 });
 
+describe('astro:config/client in a client script', () => {
+	/** @type {import('./test-utils').Fixture} */
+	let fixture;
+
+	describe('when build', () => {
+		before(async () => {
+			fixture = await loadFixture({
+				root: './fixtures/astro-manifest-client-script/',
+				adapter: testAdapter(),
+				output: 'server',
+			});
+		});
+
+		it('should build without errors when astro:config/client is used in a client script', async () => {
+			const error = await fixture.build().catch((err) => err);
+			assert.equal(error, undefined, `Build failed with: ${error?.message}`);
+		});
+	});
+});
+
 describe('astro:config/server', () => {
 	/** @type {import('./test-utils').Fixture} */
 	let fixture;

